### PR TITLE
[DNM] mgr/volumes: module for FS volumes and subvolumes

### PR DESCRIPTION
**Pull request by jcsp**
This is a big branch with progress+poolsets modules too: have a look at the single mgr/volumes commit for the interesting part.

This is currently leaving ceph_volume_client.py in its current location and re-using it, but we can later move that into the module, and replace ceph_volume_client with a stub that calls out to the mgr.

The libcephfs changes on this branch are also on a separate PR https://github.com/ceph/ceph/pull/23530

- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/pybind/mgr/poolsets/module.py [hunk 1]
```diff
@@ -0,0 +1,1408 @@
...
+ - "raw pgs": pg count after applying replication, i.e. the real resource
+              consumption of a pool.
+ - "grow/shrink" - increase/decrease the pg_num in a pool
+ - "crush subtree" - non-overlapping domains in crush hierarchy: used as
+                     units of resource management.
+"""
+
+
+# We will aim to have this many PGs per OSD
+TARGET_PGS_PER_OSD = 100
```
Comment (liewegas) [inline src/pybind/mgr/poolsets/module.py]: We have a mon_max_pg_per_osd.  Let's add a mon_target_pg_per_osd option?

(We could also s/pg/pgs/ in those)

File: src/pybind/mgr/poolsets/module.py [hunk 2]
```diff
@@ -0,0 +1,1408 @@
...
+
+# We will (if we have to) go up to this many PGs per OSD in order
+# to satisfy poolset creations.
+MAX_PGS_PER_OSD = 150
```
Comment (liewegas) [inline src/pybind/mgr/poolsets/module.py]: use existing mon_max_pg_per_osd

**Comment by batrick** [comment]
@jcsp can you rebase this now that #23530 is merged? Thanks!

File: src/pybind/mgr/volumes/module.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+            self.log.exception("Failed to tear down MDS daemons")
+            return -errno.EINVAL, "", str(e)
+
+        if self._volume_exists(vol_name):
+            # In case orchestrator didn't tear down MDS daemons cleanly, or
+            # there was no orchestrator, we force the daemons down.
+            r, out, err = self.mon_command({
+                'prefix': 'fs set',
+                'fs_name': vol_name,
+                'var': 'cluster_down',
```
Comment (batrick) [inline src/pybind/mgr/volumes/module.py]: s/cluster_down/joinable/
s/true/false/

File: src/pybind/mgr/volumes/module.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+                'val': 'true'
+            })
+            if r != 0:
+                return r, out, err
+
+            for mds_name in self._volume_get_mds_daemon_names(vol_name):
+                r, out, err = self.mon_command({
+                    'prefix': 'mds fail',
+                    'role_or_gid': mds_name})
```
Comment (batrick) [inline src/pybind/mgr/volumes/module.py]: Since you have the information, I think it'd be cleaner to use `fscid:rank` instead.

**Comment by jcsp** [comment]
@batrick this one will depend on https://github.com/ceph/ceph/pull/24178

**Comment by liewegas** [comment]
see https://github.com/ceph/ceph/pull/25272

**Pull request closed without merging**